### PR TITLE
chore: Add config for enabling SMJ with join condition

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion-comet/issues/901

## Rationale for this change

There are known correctness and performance issues with our support for SMJ with join condtion, so we should disable it by default.

## What changes are included in this PR?

New config.

### CODE REVIEW

**Comment by comphead** [comment]
Thanks, related to https://github.com/apache/datafusion/issues/9846

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 1]
```diff
@@ -145,6 +145,12 @@ object CometConf extends ShimCometConf {
   val COMET_EXEC_TAKE_ORDERED_AND_PROJECT_ENABLED: ConfigEntry[Boolean] =
     createExecEnabledConfig("takeOrderedAndProject", defaultValue = true)
 
+  val COMET_EXEC_SORT_MERGE_JOIN_WITH_JOIN_FILTER_ENABLED: ConfigEntry[Boolean] =
+    conf("spark.comet.exec.sortMergeJoinWithJoinFilter.enabled")
+      .doc("Experimental support for SMJ with join condition")
```
Comment (comphead) [inline common/src/main/scala/org/apache/comet/CometConf.scala:150]:
```suggestion
      .doc("Experimental support for Sort Merge Join with filter")
```

File: docs/source/user-guide/configs.md [hunk 2]
```diff
@@ -54,6 +54,7 @@ Comet provides the following configuration settings.
 | spark.comet.exec.shuffle.enabled | Whether to enable Comet native shuffle. Note that this requires setting 'spark.shuffle.manager' to 'org.apache.spark.sql.comet.execution.shuffle.CometShuffleManager'. 'spark.shuffle.manager' must be set before starting the Spark application and cannot be changed during the application. | true |
 | spark.comet.exec.sort.enabled | Whether to enable sort by default. | true |
 | spark.comet.exec.sortMergeJoin.enabled | Whether to enable sortMergeJoin by default. | true |
+| spark.comet.exec.sortMergeJoinWithJoinFilter.enabled | Experimental support for SMJ with join condition | false |
```
Comment (comphead) [inline docs/source/user-guide/configs.md:57]:
```suggestion
| spark.comet.exec.sortMergeJoinWithJoinFilter.enabled | Experimental support for Sort Merge Join with filter | false |
```

**Review by comphead — APPROVED** [review]
lgtm thanks @andygrove

**Review by viirya — APPROVED** [review]

File: spark/src/test/scala/org/apache/spark/sql/comet/CometPlanStabilitySuite.scala [hunk 3]
```diff
@@ -262,6 +262,7 @@ trait CometPlanStabilitySuite extends DisableAdaptiveExecutionSuite with TPCDSBa
       CometConf.COMET_EXEC_ENABLED.key -> "true",
       CometConf.COMET_DPP_FALLBACK_ENABLED.key -> "false",
       CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true",
+      CometConf.COMET_EXEC_SORT_MERGE_JOIN_WITH_JOIN_FILTER_ENABLED.key -> "true",
```
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/spark/sql/comet/CometPlanStabilitySuite.scala:265]: @andygrove Do you remember by any chance why `COMET_EXEC_SORT_MERGE_JOIN_WITH_JOIN_FILTER_ENABLED` is enabled only for `testQuery()` but not for `createSparkSession()`?

Comment (kazuyukitanimura) [reply spark/src/test/scala/org/apache/spark/sql/comet/CometPlanStabilitySuite.scala:265]: Maybe this one? https://github.com/apache/datafusion-comet/issues/901

**Pull request merged**